File: client_integration/ios/RayhunterClient/Sources/RayhunterClient/RayhunterAlertClient.swift
import Foundation
import Alamofire

/**
 * Mobile client for Rayhunter Enhanced attack detection and environment monitoring.
 * 
 * This client provides a read-only interface for mobile users to:
 * - Monitor their environment for cellular attacks
 * - Get real-time attack alerts and notifications
 * - Access forensic evidence for analysis
 * - Check device status and recording information
 * 
 * IMPORTANT: This is a read-only client - mobile users receive information
 * about their environment, they do not submit data to the device.
 */
public class RayhunterAlertClient {
    
    private let baseURL: String
    private let session: Session
    
    /**
     * Create a new RayhunterAlertClient instance.
     * 
     * - Parameter baseURL: The base URL of the Rayhunter device (e.g., "http://192.168.1.1:8080")
     * - Parameter timeout: Request timeout in seconds (default: 30)
     */
    public init(baseURL: String, timeout: TimeInterval = 30) {
        self.baseURL = baseURL
        
        let configuration = URLSessionConfiguration.default
        configuration.timeoutIntervalForRequest = timeout
        configuration.timeoutIntervalForResource = timeout
        
        self.session = Session(configuration: configuration)
    }
    
    // MARK: - Recording Information
    
    /**
     * Get the current recording manifest.
     * 
     * - Parameter completion: Completion handler with result
     */
    public func getManifest(completion: @escaping (Result<Manifest, RayhunterError>) -> Void) {
        let url = "\(baseURL)/api/qmdl-manifest"
        
        session.request(url)
            .validate()
            .responseDecodable(of: Manifest.self) { response in
                switch response.result {
                case .success(let manifest):
                    completion(.success(manifest))
                case .failure(let error):
                    completion(.failure(RayhunterError.networkError("Failed to get manifest", error)))
                }
            }
    }
    
    // MARK: - System Monitoring
    
    /**
     * Get system statistics.
     * 
     * - Parameter completion: Completion handler with result
     */
    public func getSystemStats(completion: @escaping (Result<SystemStats, RayhunterError>) -> Void) {
        let url = "\(baseURL)/api/system-stats"
        
        session.request(url)
            .validate()
            .responseDecodable(of: SystemStats.self) { response in
                switch response.result {
                case .success(let stats):
                    completion(.success(stats))
                case .failure(let error):
                    completion(.failure(RayhunterError.networkError("Failed to get system stats", error)))
                }
            }
    }
    
    // MARK: - Analysis Information
    
    /**
     * Get analysis status.
     * 
     * - Parameter completion: Completion handler with result
     */
    public func getAnalysisStatus(completion: @escaping (Result<AnalysisStatus, RayhunterError>) -> Void) {
        let url = "\(baseURL)/api/analysis"
        
        session.request(url)
            .validate()
            .responseDecodable(of: AnalysisStatus.self) { response in
                switch response.result {
                case .success(let status):
                    completion(.success(status))
                case .failure(let error):
                    completion(.failure(RayhunterError.networkError("Failed to get analysis status", error)))
                }
            }
    }
    
    /**
     * Get analysis report for a recording (contains attack alerts).
     * 
     * - Parameter recordingId: The recording ID
     * - Parameter completion: Completion handler with result
     */
    public func getAnalysisReport(recordingId: String, completion: @escaping (Result<AnalysisReport, RayhunterError>) -> Void) {
        let url = "\(baseURL)/api/analysis-report/\(recordingId)"
        
        session.request(url)
            .validate()
            .responseDecodable(of: AnalysisReport.self) { response in
                switch response.result {
                case .success(let report):
                    completion(.success(report))
                case .failure(let error):
                    completion(.failure(RayhunterError.networkError("Failed to get analysis report", error)))
                }
            }
    }
    
    /**
     * Get attack alerts for a recording (convenience method).
     * 
     * - Parameter recordingId: The recording ID
     * - Parameter completion: Completion handler with result
     */
    public func getAttackAlerts(recordingId: String, completion: @escaping (Result<AnalysisReport, RayhunterError>) -> Void) {
        getAnalysisReport(recordingId: recordingId, completion: completion)
    }
    
    /**
     * Check if a recording has any attack alerts.
     * 
     * - Parameter recordingId: The recording ID
     * - Parameter completion: Completion handler with result
     */
    public func hasAttackAlerts(recordingId: String, completion: @escaping (Result<Bool, RayhunterError>) -> Void) {
        getAnalysisReport(recordingId: recordingId) { result in
            switch result {
            case .success(let report):
                completion(.success(report.statistics.numWarnings > 0))
            case .failure(let error):
                completion(.failure(error))
            }
        }
    }
    
    /**
     * Get the count of attack alerts for a recording.
     * 
     * - Parameter recordingId: The recording ID
     * - Parameter completion: Completion handler with result
     */
    public func getAttackAlertCount(recordingId: String, completion: @escaping (Result<Int, RayhunterError>) -> Void) {
        getAnalysisReport(recordingId: recordingId) { result in
            switch result {
            case .success(let report):
                completion(.success(report.statistics.numWarnings))
            case .failure(let error):
                completion(.failure(error))
            }
        }
    }
    
    /**
     * Get the latest attack alerts from all recordings.
     * 
     * - Parameter completion: Completion handler with result
     */
    public func getLatestAttackAlerts(completion: @escaping (Result<[AttackAlertSummary], RayhunterError>) -> Void) {
        getManifest { manifestResult in
            switch manifestResult {
            case .success(let manifest):
                let group = DispatchGroup()
                var alerts: [AttackAlertSummary] = []
                let queue = DispatchQueue(label: "com.rayhunter.alerts", attributes: .concurrent)
                
                for entry in manifest.entries {
                    group.enter()
                    self.getAnalysisReport(recordingId: entry.name) { reportResult in
                        queue.async {
                            if case .success(let report) = reportResult, report.statistics.numWarnings > 0 {
                                let alert = AttackAlertSummary(
                                    recordingId: entry.name,
                                    startTime: entry.startTime,
                                    attackCount: report.statistics.numWarnings,
                                    lastMessageTime: entry.lastMessageTime
                                )
                                alerts.append(alert)
                            }
                            group.leave()
                        }
                    }
                }
                
                group.notify(queue: .main) {
                    completion(.success(alerts))
                }
                
            case .failure(let error):
                completion(.failure(error))
            }
        }
    }
    
    // MARK: - Evidence Downloads
    
    /**
     * Download evidence file for forensic analysis.
     * 
     * - Parameter endpoint: The API endpoint (e.g., "/api/pcap/123.pcapng")
     * - Parameter outputURL: The file URL to save the download to
     * - Parameter completion: Completion handler with result
     */
    public func downloadEvidenceFile(endpoint: String, outputURL: URL, completion: @escaping (Result<Void, RayhunterError>) -> Void) {
        let url = "\(baseURL)\(endpoint)"
        
        session.download(url) { _, _ in
            return (outputURL, [.removePreviousFile, .createIntermediateDirectories])
        }
        .validate()
        .responseData { response in
            switch response.result {
            case .success:
                completion(.success(()))
            case .failure(let error):
                completion(.failure(RayhunterError.networkError("Failed to download evidence file", error)))
            }
        }
    }
    
    /**
     * Download PCAP evidence file for a recording.
     * 
     * - Parameter recordingId: The recording ID
     * - Parameter outputURL: The file URL to save the PCAP to
     * - Parameter completion: Completion handler with result
     */
    public func downloadPCAPEvidence(recordingId: String, outputURL: URL, completion: @escaping (Result<Void, RayhunterError>) -> Void) {
        downloadEvidenceFile(endpoint: "/api/pcap/\(recordingId).pcapng", outputURL: outputURL, completion: completion)
    }
    
    /**
     * Download QMDL evidence file for a recording.
     * 
     * - Parameter recordingId: The recording ID
     * - Parameter outputURL: The file URL to save the QMDL to
     * - Parameter completion: Completion handler with result
     */
    public func downloadQMDLEvidence(recordingId: String, outputURL: URL, completion: @escaping (Result<Void, RayhunterError>) -> Void) {
        downloadEvidenceFile(endpoint: "/api/qmdl/\(recordingId).qmdl", outputURL: outputURL, completion: completion)
    }
    
    /**
     * Download ZIP evidence archive for a recording.
     * 
     * - Parameter recordingId: The recording ID
     * - Parameter outputURL: The file URL to save the ZIP to
     * - Parameter completion: Completion handler with result
     */
    public func downloadZIPEvidence(recordingId: String, outputURL: URL, completion: @escaping (Result<Void, RayhunterError>) -> Void) {
        downloadEvidenceFile(endpoint: "/api/zip/\(recordingId).zip", outputURL: outputURL, completion: completion)
    }
    
    // MARK: - Configuration
    
    /**
     * Get device configuration.
     * 
     * - Parameter completion: Completion handler with result
     */
    public func getConfig(completion: @escaping (Result<Config, RayhunterError>) -> Void) {
        let url = "\(baseURL)/api/config"
        
        session.request(url)
            .validate()
            .responseDecodable(of: Config.self) { response in
                switch response.result {
                case .success(let config):
                    completion(.success(config))
                case .failure(let error):
                    completion(.failure(RayhunterError.networkError("Failed to get config", error)))
                }
            }
    }
    
    /**
     * Get the base URL of this client.
     * 
     * - Returns: The base URL
     */
    public func getBaseURL() -> String {
        return baseURL
    }
}

/**
 * Summary of attack alerts for a recording.
 */
public struct AttackAlertSummary {
    public let recordingId: String
    public let startTime: String
    public let attackCount: Int
    public let lastMessageTime: String?
    
    public init(recordingId: String, startTime: String, attackCount: Int, lastMessageTime: String?) {
        self.recordingId = recordingId
        self.startTime = startTime
        self.attackCount = attackCount
        self.lastMessageTime = lastMessageTime
    }
} 

File: client_integration/android/src/main/java/com/rayhunter/client/RayhunterAlertClient.kt
package com.rayhunter.client

import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.rayhunter.client.api.RayhunterAlertApi
import com.rayhunter.client.models.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.ResponseBody
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.io.File
import java.io.FileOutputStream
import java.util.concurrent.TimeUnit

/**
 * Mobile client for Rayhunter Enhanced attack detection and environment monitoring.
 * 
 * This client provides a read-only interface for mobile users to:
 * - Monitor their environment for cellular attacks
 * - Get real-time attack alerts and notifications
 * - Access forensic evidence for analysis
 * - Check device status and recording information
 * 
 * IMPORTANT: This is a read-only client - mobile users receive information
 * about their environment, they do not submit data to the device.
 */
class RayhunterAlertClient private constructor(
    private val api: RayhunterAlertApi,
    private val baseUrl: String
) {
    
    companion object {
        private const val DEFAULT_TIMEOUT = 30L
        
        /**
         * Create a new RayhunterAlertClient instance.
         * 
         * @param baseUrl The base URL of the Rayhunter device (e.g., "http://192.168.1.1:8080")
         * @param timeoutSeconds Request timeout in seconds (default: 30)
         * @return Configured RayhunterAlertClient instance
         */
        fun create(baseUrl: String, timeoutSeconds: Long = DEFAULT_TIMEOUT): RayhunterAlertClient {
            val gson = GsonBuilder()
                .setDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'")
                .create()
            
            val client = OkHttpClient.Builder()
                .connectTimeout(timeoutSeconds, TimeUnit.SECONDS)
                .readTimeout(timeoutSeconds, TimeUnit.SECONDS)
                .writeTimeout(timeoutSeconds, TimeUnit.SECONDS)
                .build()
            
            val retrofit = Retrofit.Builder()
                .baseUrl(baseUrl)
                .client(client)
                .addConverterFactory(GsonConverterFactory.create(gson))
                .build()
            
            val api = retrofit.create(RayhunterAlertApi::class.java)
            return RayhunterAlertClient(api, baseUrl)
        }
    }
    
    /**
     * Get the current recording manifest.
     * 
     * @return Result with manifest data
     */
    suspend fun getManifest(): Result<Manifest> = withContext(Dispatchers.IO) {
        try {
            val response = api.getManifest()
            if (response.isSuccessful) {
                Result.success(response.body() ?: throw RayhunterException("Empty manifest response"))
            } else {
                Result.failure(RayhunterException("Failed to get manifest: ${response.code()}"))
            }
        } catch (e: Exception) {
            Result.failure(RayhunterException("Network error getting manifest", e))
        }
    }
    
    /**
     * Get system statistics.
     * 
     * @return Result with system stats
     */
    suspend fun getSystemStats(): Result<SystemStats> = withContext(Dispatchers.IO) {
        try {
            val response = api.getSystemStats()
            if (response.isSuccessful) {
                Result.success(response.body() ?: throw RayhunterException("Empty system stats response"))
            } else {
                Result.failure(RayhunterException("Failed to get system stats: ${response.code()}"))
            }
        } catch (e: Exception) {
            Result.failure(RayhunterException("Network error getting system stats", e))
        }
    }
    
    /**
     * Get analysis status.
     * 
     * @return Result with analysis status
     */
    suspend fun getAnalysisStatus(): Result<AnalysisStatus> = withContext(Dispatchers.IO) {
        try {
            val response = api.getAnalysisStatus()
            if (response.isSuccessful) {
                Result.success(response.body() ?: throw RayhunterException("Empty analysis status response"))
            } else {
                Result.failure(RayhunterException("Failed to get analysis status: ${response.code()}"))
            }
        } catch (e: Exception) {
            Result.failure(RayhunterException("Network error getting analysis status", e))
        }
    }
    
    /**
     * Get analysis report for a recording (contains attack alerts).
     * 
     * @param recordingId The recording ID
     * @return Result with analysis report containing attack alerts
     */
    suspend fun getAnalysisReport(recordingId: String): Result<AnalysisReport> = withContext(Dispatchers.IO) {
        try {
            val response = api.getAnalysisReport(recordingId)
            if (response.isSuccessful) {
                Result.success(response.body() ?: throw RayhunterException("Empty analysis report response"))
            } else {
                Result.failure(RayhunterException("Failed to get analysis report: ${response.code()}"))
            }
        } catch (e: Exception) {
            Result.failure(RayhunterException("Network error getting analysis report", e))
        }
    }
    
    /**
     * Get attack alerts for a recording (convenience method).
     * 
     * @param recordingId The recording ID
     * @return Result with analysis report containing attack alerts
     */
    suspend fun getAttackAlerts(recordingId: String): Result<AnalysisReport> {
        return getAnalysisReport(recordingId)
    }
    
    /**
     * Check if a recording has any attack alerts.
     * 
     * @param recordingId The recording ID
     * @return Result with boolean indicating if attacks were detected
     */
    suspend fun hasAttackAlerts(recordingId: String): Result<Boolean> = withContext(Dispatchers.IO) {
        try {
            val report = getAnalysisReport(recordingId).getOrThrow()
            Result.success(report.statistics.numWarnings > 0)
        } catch (e: Exception) {
            Result.failure(RayhunterException("Error checking for attack alerts", e))
        }
    }
    
    /**
     * Get the count of attack alerts for a recording.
     * 
     * @param recordingId The recording ID
     * @return Result with number of attack alerts
     */
    suspend fun getAttackAlertCount(recordingId: String): Result<Int> = withContext(Dispatchers.IO) {
        try {
            val report = getAnalysisReport(recordingId).getOrThrow()
            Result.success(report.statistics.numWarnings)
        } catch (e: Exception) {
            Result.failure(RayhunterException("Error getting attack alert count", e))
        }
    }
    
    /**
     * Get the latest attack alerts from all recordings.
     * 
     * @return Result with list of recordings that have attack alerts
     */
    suspend fun getLatestAttackAlerts(): Result<List<AttackAlertSummary>> = withContext(Dispatchers.IO) {
        try {
            val manifest = getManifest().getOrThrow()
            val alerts = mutableListOf<AttackAlertSummary>()
            
            for (entry in manifest.entries) {
                val report = getAnalysisReport(entry.name).getOrNull()
                if (report != null && report.statistics.numWarnings > 0) {
                    alerts.add(AttackAlertSummary(
                        recordingId = entry.name,
                        startTime = entry.startTime,
                        attackCount = report.statistics.numWarnings,
                        lastMessageTime = entry.lastMessageTime
                    ))
                }
            }
            
            Result.success(alerts)
        } catch (e: Exception) {
            Result.failure(RayhunterException("Error getting latest attack alerts", e))
        }
    }
    
    /**
     * Download evidence files for forensic analysis.
     * 
     * @param endpoint The API endpoint (e.g., "/api/pcap/123.pcapng")
     * @param outputFile The file to save the download to
     * @return Result indicating success or failure
     */
    suspend fun downloadEvidenceFile(endpoint: String, outputFile: File): Result<Unit> = withContext(Dispatchers.IO) {
        try {
            val response = api.downloadFile(endpoint)
            if (response.isSuccessful) {
                val body = response.body() ?: throw RayhunterException("Empty download response")
                val inputStream = body.byteStream()
                val outputStream = FileOutputStream(outputFile)
                
                inputStream.use { input ->
                    outputStream.use { output ->
                        input.copyTo(output)
                    }
                }
                Result.success(Unit)
            } else {
                Result.failure(RayhunterException("Failed to download evidence file: ${response.code()}"))
            }
        } catch (e: Exception) {
            Result.failure(RayhunterException("Network error downloading evidence file", e))
        }
    }
    
    /**
     * Download PCAP evidence file for a recording.
     * 
     * @param recordingId The recording ID
     * @param outputFile The file to save the PCAP to
     * @return Result indicating success or failure
     */
    suspend fun downloadPCAPEvidence(recordingId: String, outputFile: File): Result<Unit> {
        return downloadEvidenceFile("/api/pcap/$recordingId.pcapng", outputFile)
    }
    
    /**
     * Download QMDL evidence file for a recording.
     * 
     * @param recordingId The recording ID
     * @param outputFile The file to save the QMDL to
     * @return Result indicating success or failure
     */
    suspend fun downloadQMDLEvidence(recordingId: String, outputFile: File): Result<Unit> {
        return downloadEvidenceFile("/api/qmdl/$recordingId.qmdl", outputFile)
    }
    
    /**
     * Download ZIP evidence archive for a recording.
     * 
     * @param recordingId The recording ID
     * @param outputFile The file to save the ZIP to
     * @return Result indicating success or failure
     */
    suspend fun downloadZIPEvidence(recordingId: String, outputFile: File): Result<Unit> {
        return downloadEvidenceFile("/api/zip/$recordingId.zip", outputFile)
    }
    
    /**
     * Get device configuration.
     * 
     * @return Result with configuration data
     */
    suspend fun getConfig(): Result<Config> = withContext(Dispatchers.IO) {
        try {
            val response = api.getConfig()
            if (response.isSuccessful) {
                Result.success(response.body() ?: throw RayhunterException("Empty config response"))
            } else {
                Result.failure(RayhunterException("Failed to get config: ${response.code()}"))
            }
        } catch (e: Exception) {
            Result.failure(RayhunterException("Network error getting config", e))
        }
    }
    
    /**
     * Get the base URL of this client.
     * 
     * @return The base URL
     */
    fun getBaseUrl(): String = baseUrl
}

/**
 * Summary of attack alerts for a recording.
 */
data class AttackAlertSummary(
    val recordingId: String,
    val startTime: String,
    val attackCount: Int,
    val lastMessageTime: String?
) 